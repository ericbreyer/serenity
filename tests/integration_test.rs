#![cfg(test)]

use anyhow::Result;
use insta::{assert_debug_snapshot, assert_snapshot};
use test_case::test_case;

#[test_case("trivial.ser", 1; "trivial")]
#[test_case("prime_sieve.ser", 8181807856294299570; "prime_sieve")]
#[test_case("interfaces.ser", 12; "interfaces")]
#[test_case("babbage.ser", 25264; "babbage")]
#[test_case("linkedlist.ser", 55; "linkedlist")]
#[test_case("prime_conspiricy.ser", 7942686168; "prime_conspiricy")]
#[test_case("highly_composites.ser", 355168; "highly_composites")]
pub fn test_file_run(file : &str, ecode: i64) -> Result<()> {

    let file = format!("tests/{}", file);

    let mut out = Vec::new();
    let code = serenity::run_file(&file, &mut out)?;

    assert!(code == ecode, "Expected: {}, Got: {}", ecode, code);    
    
    Ok(())
}

#[test_case("trivial.ser"; "trivial")]
#[test_case("prime_sieve.ser"; "prime_sieve")]
#[test_case("interfaces.ser"; "interfaces")]
#[test_case("babbage.ser"; "babbage")]
#[test_case("linkedlist.ser"; "linkedlist")]
#[test_case("prime_conspiricy.ser"; "prime_conspiricy")]
#[test_case("highly_composites.ser"; "highly_composites")]
pub fn test_file_scan(file : &str) -> Result<()> {
    let file = format!("tests/{}", file);
    let tokens = serenity::scan(&file)?;

    let mut settings = insta::Settings::clone_current();
    settings.set_snapshot_suffix(file);
    let _guard = settings.bind_to_scope();

    assert_debug_snapshot!(tokens);
    Ok(())
}

#[test_case("trivial.ser"; "trivial")]
#[test_case("prime_sieve.ser"; "prime_sieve")]
#[test_case("interfaces.ser"; "interfaces")]
#[test_case("babbage.ser"; "babbage")]
#[test_case("linkedlist.ser"; "linkedlist")]
#[test_case("prime_conspiricy.ser"; "prime_conspiricy")]
#[test_case("highly_composites.ser"; "highly_composites")]
pub fn test_file_parse(file : &str) -> Result<()> {
    let file = format!("tests/{}", file);
    let ast = serenity::parse(&file)?;

    let mut settings = insta::Settings::clone_current();
    settings.set_snapshot_suffix(file);
    let _guard = settings.bind_to_scope();

    assert_snapshot!(ast);
    Ok(())
}

#[test_case("trivial.ser"; "trivial")]
#[test_case("prime_sieve.ser"; "prime_sieve")]
#[test_case("interfaces.ser"; "interfaces")]
#[test_case("babbage.ser"; "babbage")]
#[test_case("linkedlist.ser"; "linkedlist")]
#[test_case("prime_conspiricy.ser"; "prime_conspiricy")]
#[test_case("highly_composites.ser"; "highly_composites")]
pub fn test_file_compile(file : &str) -> Result<()> {
    let file = format!("tests/{}", file);
    let ir = serenity::compile(&file)?;

    let mut settings = insta::Settings::clone_current();
    settings.set_snapshot_suffix(file);
    let _guard = settings.bind_to_scope();

    assert_snapshot!(ir);
    Ok(())
}

