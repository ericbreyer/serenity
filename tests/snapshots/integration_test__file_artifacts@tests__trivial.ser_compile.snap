---
source: tests/integration_test.rs
expression: result
---
; ModuleID = 'main'
source_filename = "main"

@"7250923139693181361_next" = global { ptr } { ptr @"7250923139693181361_next_fn" }
@"15292438991726710161_next" = global { ptr } { ptr @"15292438991726710161_next_fn" }
@nat_gen_generator_vtable = global { { ptr } } zeroinitializer
@nat_gen_impl_generator = global { ptr } { ptr @nat_gen_impl_generator_fn }
@main = global { ptr } { ptr @main_fn }

declare i8 @printf(ptr, ...)

declare ptr @malloc(i64)

declare i8 @free(ptr)

define { ptr, ptr } @"7250923139693181361_next_fn"(ptr %0) {
entry:
  %closure = alloca { ptr, ptr }, align 8
  %self = alloca ptr, align 8
  store ptr %0, ptr %self, align 8
  %self1 = load ptr, ptr %self, align 8
  %upval = getelementptr inbounds { ptr, ptr }, ptr %closure, i32 0, i32 0
  store ptr %self1, ptr %upval, align 8
  %upval2 = getelementptr inbounds { ptr, ptr }, ptr %closure, i32 0, i32 1
  store ptr @"7250923139693181361_next_inner_fn", ptr %upval2, align 8
  %closure3 = load { ptr, ptr }, ptr %closure, align 8
  ret { ptr, ptr } %closure3
}

define i64 @"7250923139693181361_next_inner_fn"(ptr %0) {
entry:
  %calltmp11 = alloca i64, align 8
  %calltmp7 = alloca { ptr, ptr }, align 8
  %self = alloca ptr, align 8
  store ptr %0, ptr %self, align 8
  %rval = load ptr, ptr %self, align 8
  %field = getelementptr inbounds { ptr, ptr }, ptr %rval, i32 0, i32 1
  %rval1 = load ptr, ptr %field, align 8
  %field2 = getelementptr inbounds { { ptr } }, ptr %rval1, i32 0, i32 0
  %rval3 = load { ptr }, ptr %field2, align 8
  %rval4 = load ptr, ptr %self, align 8
  %field5 = getelementptr inbounds { ptr, ptr }, ptr %rval4, i32 0, i32 0
  %rval6 = load ptr, ptr %field5, align 8
  %fnptr = extractvalue { ptr } %rval3, 0
  %calltmp = call { ptr, ptr } %fnptr(ptr %rval6)
  store { ptr, ptr } %calltmp, ptr %calltmp7, align 8
  %rval8 = load { ptr, ptr }, ptr %calltmp7, align 8
  %fnptr9 = extractvalue { ptr, ptr } %rval8, 1
  %upval = extractvalue { ptr, ptr } %rval8, 0
  %calltmp10 = call i64 %fnptr9(ptr %upval)
  store i64 %calltmp10, ptr %calltmp11, align 4
  %rval12 = load i64, ptr %calltmp11, align 4
  ret i64 %rval12
}

define { ptr, ptr } @"15292438991726710161_next_fn"(ptr %0) {
entry:
  %closure7 = alloca { ptr, ptr }, align 8
  %self5 = alloca ptr, align 8
  %closure = alloca { ptr, ptr }, align 8
  %self = alloca ptr, align 8
  store ptr %0, ptr %self, align 8
  %self1 = load ptr, ptr %self, align 8
  %upval = getelementptr inbounds { ptr, ptr }, ptr %closure, i32 0, i32 0
  store ptr %self1, ptr %upval, align 8
  %upval2 = getelementptr inbounds { ptr, ptr }, ptr %closure, i32 0, i32 1
  store ptr @"15292438991726710161_next_inner_fn", ptr %upval2, align 8
  %closure3 = load { ptr, ptr }, ptr %closure, align 8
  ret { ptr, ptr } %closure3

entry4:                                           ; No predecessors!
  store ptr %0, ptr %self5, align 8
  %self6 = load ptr, ptr %self5, align 8
  %upval8 = getelementptr inbounds { ptr, ptr }, ptr %closure7, i32 0, i32 0
  store ptr %self6, ptr %upval8, align 8
  %upval9 = getelementptr inbounds { ptr, ptr }, ptr %closure7, i32 0, i32 1
  store ptr @"15292438991726710161_next_inner_fn", ptr %upval9, align 8
  %closure10 = load { ptr, ptr }, ptr %closure7, align 8
  ret { ptr, ptr } %closure10
}

define i64 @"15292438991726710161_next_inner_fn"(ptr %0) {
entry:
  %i = alloca i64, align 8
  %self = alloca ptr, align 8
  store ptr %0, ptr %self, align 8
  %rval = load ptr, ptr %self, align 8
  %field = getelementptr inbounds { i64 }, ptr %rval, i32 0, i32 0
  %rval1 = load i64, ptr %field, align 4
  store i64 %rval1, ptr %i, align 4
  %rval2 = load i64, ptr %i, align 4
  %addtmp = add i64 %rval2, 1
  %rval3 = load ptr, ptr %self, align 8
  %field4 = getelementptr inbounds { i64 }, ptr %rval3, i32 0, i32 0
  store i64 %addtmp, ptr %field4, align 4
  %rval5 = load i64, ptr %i, align 4
  ret i64 %rval5
}

define { ptr, ptr } @nat_gen_impl_generator_fn(ptr %0) {
entry:
  %struct_init = alloca { ptr, ptr }, align 8
  %self = alloca ptr, align 8
  store ptr %0, ptr %self, align 8
  %rval = load { ptr }, ptr @"15292438991726710161_next", align 8
  store { ptr } %rval, ptr @nat_gen_generator_vtable, align 8
  %field = getelementptr inbounds { ptr, ptr }, ptr %struct_init, i32 0, i32 0
  %rval1 = load ptr, ptr %self, align 8
  store ptr %rval1, ptr %field, align 8
  %field2 = getelementptr inbounds { ptr, ptr }, ptr %struct_init, i32 0, i32 1
  store ptr @nat_gen_generator_vtable, ptr %field2, align 8
  %struct_init3 = load { ptr, ptr }, ptr %struct_init, align 8
  ret { ptr, ptr } %struct_init3
}

define i64 @main_fn() {
entry:
  %calltmp21 = alloca i64, align 8
  %calltmp16 = alloca { ptr, ptr }, align 8
  %calltmp11 = alloca i64, align 8
  %calltmp7 = alloca { ptr, ptr }, align 8
  %g = alloca { ptr, ptr }, align 8
  %calltmp2 = alloca { ptr, ptr }, align 8
  %gen = alloca { i64 }, align 8
  %struct_init = alloca { i64 }, align 8
  %field = getelementptr inbounds { i64 }, ptr %struct_init, i32 0, i32 0
  store i64 0, ptr %field, align 4
  %struct_init1 = load { i64 }, ptr %struct_init, align 4
  store { i64 } %struct_init1, ptr %gen, align 4
  %rval = load { ptr }, ptr @nat_gen_impl_generator, align 8
  %fnptr = extractvalue { ptr } %rval, 0
  %calltmp = call { ptr, ptr } %fnptr(ptr %gen)
  store { ptr, ptr } %calltmp, ptr %calltmp2, align 8
  %rval3 = load { ptr, ptr }, ptr %calltmp2, align 8
  store { ptr, ptr } %rval3, ptr %g, align 8
  %rval4 = load { ptr }, ptr @"7250923139693181361_next", align 8
  %fnptr5 = extractvalue { ptr } %rval4, 0
  %calltmp6 = call { ptr, ptr } %fnptr5(ptr %g)
  store { ptr, ptr } %calltmp6, ptr %calltmp7, align 8
  %rval8 = load { ptr, ptr }, ptr %calltmp7, align 8
  %fnptr9 = extractvalue { ptr, ptr } %rval8, 1
  %upval = extractvalue { ptr, ptr } %rval8, 0
  %calltmp10 = call i64 %fnptr9(ptr %upval)
  store i64 %calltmp10, ptr %calltmp11, align 4
  %rval12 = load i64, ptr %calltmp11, align 4
  %rval13 = load { ptr }, ptr @"7250923139693181361_next", align 8
  %fnptr14 = extractvalue { ptr } %rval13, 0
  %calltmp15 = call { ptr, ptr } %fnptr14(ptr %g)
  store { ptr, ptr } %calltmp15, ptr %calltmp16, align 8
  %rval17 = load { ptr, ptr }, ptr %calltmp16, align 8
  %fnptr18 = extractvalue { ptr, ptr } %rval17, 1
  %upval19 = extractvalue { ptr, ptr } %rval17, 0
  %calltmp20 = call i64 %fnptr18(ptr %upval19)
  store i64 %calltmp20, ptr %calltmp21, align 4
  %rval22 = load i64, ptr %calltmp21, align 4
  %addtmp = add i64 %rval12, %rval22
  ret i64 %addtmp
}
